File: Serenity.Script.UI/FilterPanel/FilteringTypes/LookupFiltering.cs
using System.Collections.Generic;
using System.Runtime.CompilerServices;

namespace Serenity
{
    [Imported(ObeysTypeSystem = true)]
    public class LookupFiltering : BaseEditorFiltering
    {
        public LookupFiltering()
            : base(typeof(LookupEditor))
        {
        }

        public override List<FilterOperator> GetOperators()
        {
            return AppendNullableOperators(
                new List<FilterOperator>
                {
                    FilterOperators.EQ,
                    FilterOperators.NE,
                    FilterOperators.Contains,
                    FilterOperators.StartsWith
                });
        }

        protected override bool UseEditor()
        {
            return Operator.Key == FilterOperators.EQ || 
                Operator.Key == FilterOperators.NE;
        }

        protected override bool UseIdField()
        {
            return UseEditor();
        }

        protected override string GetEditorText()
        {
            if (UseEditor())
                return editor.As<LookupEditor>().Text;

            return base.GetEditorText();
        }
    }
}

File: packages/corelib/src/ui/filtering/lookupfiltering.ts
import { nsSerenity } from "../../base";
import { LookupEditor } from "../editors/lookupeditor";
import { BaseEditorFiltering } from "./baseeditorfiltering";
import { FilterOperator, FilterOperators } from "./filteroperator";

export class LookupFiltering extends BaseEditorFiltering<LookupEditor> {
    static override [Symbol.typeInfo] = this.registerClass(nsSerenity);

    constructor() {
        super(LookupEditor);
    }

    getOperators(): FilterOperator[] {
        var ops = [{ key: FilterOperators.EQ }, { key: FilterOperators.NE }, { key: FilterOperators.contains }, { key: FilterOperators.startsWith }]
        return this.appendNullableOperators(ops);
    }

    protected override useEditor(): boolean {
        var op = this.get_operator().key;
        return op == FilterOperators.EQ || op == FilterOperators.NE;
    }

    protected override useIdField(): boolean {
        return this.useEditor();
    }

    override getEditorText(): string {
        if (this.useEditor()) {
            return this.editor.text;
        }

        return super.getEditorText();
    }
}
